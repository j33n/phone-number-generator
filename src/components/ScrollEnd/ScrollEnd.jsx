import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import arrow from '../../assets/arrow.svg';

import './ScrollEnd.scss';

class ScrollEnd extends PureComponent {
    static propTypes = {
      size: PropTypes.number,
      height: PropTypes.number,
    }

    static defaultProps = {
      size: 32,
      height: 0,
    }

    state = {
      scrollY: 0,
      topBottomScroll: false,
      bottomTopScroll: false
    }

    componentDidMount() {
      window.addEventListener('scroll', this.handleScroll);
    }

    componentDidUpdate = (prevProps, prevState) => {
      const { scrollY } = this.state;
      if (prevState.scrollY !== scrollY && scrollY > prevState.scrollY && scrollY !== 0) {
        this.setState({
          topBottomScroll: true,
          bottomTopScroll: false
        });
      }

      if (prevState.scrollY !== scrollY && scrollY < prevState.scrollY && scrollY !== 0) {
        this.setState({
          topBottomScroll: false,
          bottomTopScroll: true
        });
      }
    }

    componentWillUnmount() {
      window.removeEventListener('scroll', this.handleScroll);
    }


    scrollToEnd = () => {
      const { bottomTopScroll, topBottomScroll } = this.state;
      const { height } = this.props;

      if (topBottomScroll) {
        window.scrollTo(0, height);
      } else if (bottomTopScroll) {
        window.scrollTo(0, 0);
      } else {
        window.scrollTo(0, 0);
      }
    }

    handleScroll = () => {
      const { scrollY } = window;
      if (scrollY === 0) {
        return this.setState({
          scrollY: 0,
          topBottomScroll: false,
          bottomTopScroll: false,
        });
      }
      return this.setState({ scrollY });
    }

    render() {
      const { size } = this.props;
      const { bottomTopScroll, topBottomScroll } = this.state;
      const imageStyle = {
        width: size,
      };
      return (
        (bottomTopScroll || topBottomScroll)
            && (
            <div className="scroll" onClick={this.scrollToEnd} role="presentation">
              <img src={arrow} alt="arrow" style={imageStyle} className={`arrow-button ${bottomTopScroll && 'scroll-up'} ${topBottomScroll && 'scroll-down'}`} />
            </div>
            )
      );
    }
}

export default ScrollEnd;
