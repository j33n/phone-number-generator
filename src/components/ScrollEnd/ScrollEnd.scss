.scroll {
    position: fixed;
    background-image: linear-gradient(to right, #009493, #418280, #008281);
    top: 10%;
    right: 10px;
    padding: 5px;
    border-radius: 10px;
    height: 32px;

    &:hover {
        cursor: pointer;
        background-image: linear-gradient(to right, #009493, #0d9692, #008281);
    }

    .scroll-up {
        transform: rotate(180deg);
    }
}