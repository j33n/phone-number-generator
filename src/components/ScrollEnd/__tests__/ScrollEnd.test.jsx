import React from 'react';
import { shallow } from 'enzyme';

import ScrollEnd from '../ScrollEnd';

describe('NumberTable', () => {
  const component = <ScrollEnd size={12} height={12} />;

  jest.useFakeTimers();

  it('renders', () => {
    expect(component).toMatchSnapshot();
  });

  it('should auto set scrollY to 0 by default', () => {
    window.addEventListener = jest.fn();
    window.scrollY = 0;
    const subject = shallow(component);
    expect(window.addEventListener).toHaveBeenCalled();
    expect(subject.state('scrollY')).toEqual(0);
  });

  it('should auto set scrollY', () => {
    window.addEventListener = jest.fn();
    window.scrollY = 100;
    const subject = shallow(component);
    expect(window.addEventListener).toHaveBeenCalled();
    expect(subject.state('scrollY')).toEqual(100);
  });

  it('should identify vertical scroll', () => {
    window.scrollY = 100;
    const subject = shallow(component);
    expect(subject.state('scrollY')).toEqual(100);
    expect(subject.state('topBottomScroll')).toBeTruthy();
    expect(subject.state('bottomTopScroll')).toBeFalsy();
  });

  it('should identify horizontal scroll', () => {
    window.scrollY = -100;
    const subject = shallow(component);
    expect(subject.state('scrollY')).toEqual(-100);
    expect(subject.state('topBottomScroll')).toBeFalsy();
    expect(subject.state('bottomTopScroll')).toBeTruthy();
  });

  it('should afford scrolling to the far bottom', () => {
    window.scrollY = 100;
    window.scrollTo = jest.fn();
    const subject = shallow(component);
    subject.find('.scroll').simulate('click');
    expect(window.scrollTo).toBeCalledWith(expect.any(Number), expect.any(Number));
  });

  it('should afford scrolling to the far top', () => {
    window.scrollY = -100;
    window.scrollTo = jest.fn();
    const subject = shallow(component);
    subject.find('.scroll').simulate('click');
    expect(window.scrollTo).toBeCalledWith(expect.any(Number), expect.any(Number));
  });

  it('should default to 0, 0', () => {
    window.scrollY = 0;
    window.scrollTo = jest.fn();
    const subject = shallow(component);
    subject.instance().scrollToEnd();
    expect(window.scrollTo).toBeCalledWith(0, 0);
  });
});
