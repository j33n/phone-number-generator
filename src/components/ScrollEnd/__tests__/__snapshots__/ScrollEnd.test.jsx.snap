// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberTable renders 1`] = `
<ScrollEnd
  height={12}
  size={12}
/>
`;
